File: backend/app/schemas/workflow.py
"""
Workflow Schemas - 工作流相关数据验证模型
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from uuid import UUID
from enum import Enum

from .base import AuditSchema, SoftDeleteSchema


# ==================== Enums ====================

class TriggerTypeEnum(str, Enum):
    """触发器类型枚举"""
    # 工作项相关
    ISSUE_CREATED = "issue.created"
    ISSUE_UPDATED = "issue.updated"
    ISSUE_DELETED = "issue.deleted"
    ISSUE_ASSIGNED = "issue.assigned"
    STATE_CHANGED = "issue.state_changed"
    PRIORITY_CHANGED = "issue.priority_changed"
    DUE_SOON = "issue.due_soon"
    DUE_DATE_PASSED = "issue.due_date_passed"
    
    # 周期相关
    CYCLE_STARTED = "cycle.started"
    CYCLE_ENDED = "cycle.ended"
    
    # 评论相关
    COMMENT_ADDED = "comment.added"


class ConditionOperatorEnum(str, Enum):
    """条件操作符枚举"""
    EQUALS = "equals"
    NOT_EQUALS = "not_equals"
    CONTAINS = "contains"
    NOT_CONTAINS = "not_contains"
    IN = "in"
    NOT_IN = "not_in"
    GREATER_THAN = "greater_than"
    LESS_THAN = "less_than"
    IS_EMPTY = "is_empty"
    IS_NOT_EMPTY = "is_not_empty"


class ActionTypeEnum(str, Enum):
    """动作类型枚举"""
    # 工作项动作
    ISSUE_UPDATE = "issue.update"
    ISSUE_ASSIGN = "issue.assign"
    ISSUE_ADD_LABEL = "issue.add_label"
    ISSUE_REMOVE_LABEL = "issue.remove_label"
    ISSUE_CHANGE_STATE = "issue.change_state"
    ISSUE_SET_PRIORITY = "issue.set_priority"
    
    # 通知动作
    NOTIFICATION_CREATE = "notification.create"
    EMAIL_SEND = "email.send"
    
    # 自动化动作
    AUTOMATION_TRIGGER = "automation.trigger"


# ==================== State Transition Schema ====================

class StateTransitionBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    is_auto: bool = Field(default=False)


class StateTransitionCreate(StateTransitionBase):
    source_state_id: UUID
    target_state_id: UUID
    issue_type_id: Optional[UUID] = None


class StateTransitionUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    is_auto: Optional[bool] = None
    issue_type_id: Optional[UUID] = None


class StateTransitionResponse(AuditSchema, SoftDeleteSchema, StateTransitionBase):
    id: UUID
    source_state_id: UUID
    target_state_id: UUID
    issue_type_id: Optional[UUID] = None
    project_id: UUID
    workspace_id: UUID
    
    class Config:
        from_attributes = True


# ==================== Automation Rule Schema ====================

class TriggerSchema(BaseModel):
    """触发器配置"""
    type: TriggerTypeEnum
    # 额外配置
    state_group: Optional[str] = None  # 用于 state_changed 触发器
    priority: Optional[str] = None  # 用于 priority_changed 触发器
    days_before: Optional[int] = None  # 用于 due_soon 触发器
    field: Optional[str] = None  # 用于 updated 触发器


class ConditionSchema(BaseModel):
    """条件配置"""
    field: str  # 如 "state.group", "priority", "assignee_id"
    operator: ConditionOperatorEnum
    value: Optional[Any] = None  # 比较值


class ActionSchema(BaseModel):
    """动作配置"""
    type: ActionTypeEnum
    # 工作项动作
    field: Optional[str] = None  # 如 "assignee", "priority"
    value: Optional[Any] = None
    label: Optional[str] = None  # 用于添加/移除标签
    state_id: Optional[UUID] = None  # 用于改变状态
    # 通知动作
    message: Optional[str] = None
    subject: Optional[str] = None
    recipients: Optional[List[str]] = None


class AutomationRuleBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    is_enabled: bool = Field(default=True)


class AutomationRuleCreate(AutomationRuleBase):
    project_id: UUID
    trigger: Dict[str, Any]  # 使用 dict 而非 TriggerSchema，方便扩展
    conditions: List[Dict[str, Any]] = Field(default=[])
    actions: List[Dict[str, Any]] = Field(default=[])


class AutomationRuleUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=500)
    is_enabled: Optional[bool] = None
    trigger: Optional[Dict[str, Any]] = None
    conditions: Optional[List[Dict[str, Any]]] = None
    actions: Optional[List[Dict[str, Any]]] = None


class AutomationRuleResponse(AuditSchema, SoftDeleteSchema, AutomationRuleBase):
    id: UUID
    trigger: Dict[str, Any]
    conditions: List[Dict[str, Any]]
    actions: List[Dict[str, Any]]
    execution_count: int
    last_executed_at: Optional[str] = None
    project_id: UUID
    workspace_id: UUID
    
    class Config:
        from_attributes = True


class AutomationRuleLite(BaseModel):
    """轻量级自动化规则响应"""
    id: UUID
    name: str
    description: Optional[str] = None
    is_enabled: bool
    trigger: Dict[str, Any]
    execution_count: int
    
    class Config:
        from_attributes = True


# ==================== Automation Execution Log Schema ====================

class AutomationExecutionLogResponse(BaseModel):
    id: UUID
    rule_id: UUID
    status: str  # success, failed, skipped
    trigger_event: str
    triggered_issue_id: Optional[UUID] = None
    execution_details: Dict[str, Any]
    error_message: Optional[str] = None
    execution_time_ms: Optional[int] = None
    created_at: str
    
    class Config:
        from_attributes = True


# ==================== Automation Template Schema ====================

class AutomationTemplateResponse(BaseModel):
    id: UUID
    name: str
    description: Optional[str] = None
    category: str
    trigger: Dict[str, Any]
    conditions: List[Dict[str, Any]]
    actions: List[Dict[str, Any]]
    is_system: bool
    usage_count: int
    
    class Config:
        from_attributes = True


# ==================== Helper Functions ====================

def get_trigger_display_name(trigger_type: TriggerTypeEnum) -> str:
    """获取触发器类型的显示名称"""
    names = {
        TriggerTypeEnum.ISSUE_CREATED: "工作项创建时",
        TriggerTypeEnum.ISSUE_UPDATED: "工作项更新时",
        TriggerTypeEnum.ISSUE_DELETED: "工作项删除时",
        TriggerTypeEnum.ISSUE_ASSIGNED: "工作项分配时",
        TriggerTypeEnum.STATE_CHANGED: "状态变更时",
        TriggerTypeEnum.PRIORITY_CHANGED: "优先级变更时",
        TriggerTypeEnum.DUE_SOON: "截止日期临近时",
        TriggerTypeEnum.DUE_DATE_PASSED: "截止日期过期时",
        TriggerTypeEnum.CYCLE_STARTED: "周期开始时",
        TriggerTypeEnum.CYCLE_ENDED: "周期结束时",
        TriggerTypeEnum.COMMENT_ADDED: "添加评论时",
    }
    return names.get(trigger_type, trigger_type)


def get_action_display_name(action_type: ActionTypeEnum) -> str:
    """获取动作类型的显示名称"""
    names = {
        ActionTypeEnum.ISSUE_UPDATE: "更新工作项",
        ActionTypeEnum.ISSUE_ASSIGN: "分配工作项",
        ActionTypeEnum.ISSUE_ADD_LABEL: "添加标签",
        ActionTypeEnum.ISSUE_REMOVE_LABEL: "移除标签",
        ActionTypeEnum.ISSUE_CHANGE_STATE: "改变状态",
        ActionTypeEnum.ISSUE_SET_PRIORITY: "设置优先级",
        ActionTypeEnum.NOTIFICATION_CREATE: "创建通知",
        ActionTypeEnum.EMAIL_SEND: "发送邮件",
        ActionTypeEnum.AUTOMATION_TRIGGER: "触发自动化",
    }
    return names.get(action_type, action_type)


# ==================== Default Workflow Transitions ====================

# 默认状态转换规则：允许大部分状态之间的转换
DEFAULT_STATE_TRANSITIONS = [
    # Backlog 可以转到任何状态
    {"name": "Backlog → Todo", "source_group": "backlog", "target_group": "todo"},
    # Todo 可以转到进行中或取消
    {"name": "Todo → In Progress", "source_group": "todo", "target_group": "in_progress"},
    {"name": "Todo → Cancelled", "source_group": "todo", "target_group": "cancelled"},
    # In Progress 可以转到完成、Review 或回到 Todo
    {"name": "In Progress → In Review", "source_group": "in_progress", "target_group": "in_progress"},
    {"name": "In Progress → Done", "source_group": "in_progress", "target_group": "done"},
    {"name": "In Progress → Cancelled", "source_group": "in_progress", "target_group": "cancelled"},
    # In Review 可以转到完成或回到进行中
    {"name": "In Review → Done", "source_group": "in_progress", "target_group": "done"},
    {"name": "In Review → In Progress", "source_group": "in_progress", "target_group": "in_progress"},
    # Done 和 Cancelled 是终态，可以重新打开
    {"name": "Done → Todo", "source_group": "done", "target_group": "todo"},
    {"name": "Cancelled → Todo", "source_group": "cancelled", "target_group": "todo"},
]

File: frontend/src/types/workflow.ts
/**
 * Workflow Types - 工作流和自动化类型定义
 */

// ==================== Enums ====================

export enum TriggerTypeEnum {
  // 工作项相关
  ISSUE_CREATED = 'issue.created',
  ISSUE_UPDATED = 'issue.updated',
  ISSUE_DELETED = 'issue.deleted',
  ISSUE_ASSIGNED = 'issue.assigned',
  STATE_CHANGED = 'issue.state_changed',
  DUE_SOON = 'issue.due_soon',
  DUE_DATE_PASSED = 'issue.due_date_passed',
  
  // 周期相关
  CYCLE_STARTED = 'cycle.started',
  CYCLE_ENDED = 'cycle.ended',
  
  // 评论相关
  COMMENT_ADDED = 'comment.added',

  // 定时触发器
  SCHEDULED = 'scheduled'
}

export const TriggerTypeOptions = [
  { value: 'issue.created', label: '工作项创建时', icon: '➕' },
  { value: 'issue.updated', label: '工作项更新时', icon: '✏️' },
  { value: 'issue.state_changed', label: '状态变更时', icon: '🔄' },
  { value: 'issue.assigned', label: '分配人变更时', icon: '👤' },
  { value: 'issue.due_soon', label: '截止日期临近时', icon: '⏰' },
  { value: 'issue.due_date_passed', label: '截止日期过期时', icon: '⚠️' },
  { value: 'cycle.started', label: '周期开始时', icon: '▶️' },
  { value: 'cycle.ended', label: '周期结束时', icon: '⏹️' },
  { value: 'comment.added', label: '添加评论时', icon: '💬' },
  { value: 'scheduled', label: '定时执行', icon: '⏱️' },
]

export const StateGroupOptions = [
  { value: 'backlog', label: '待处理 (Backlog)', color: '#6B7280' },
  { value: 'unstarted', label: '未开始 (Todo)', color: '#3B82F6' },
  { value: 'started', label: '进行中 (In Progress)', color: '#8B5CF6' },
  { value: 'completed', label: '已完成 (Done)', color: '#10B981' },
  { value: 'cancelled', label: '已取消 (Cancelled)', color: '#EF4444' },
]

export const PriorityOptions = [
  { value: 'urgent', label: '紧急 (Urgent)', color: '#EF4444' },
  { value: 'high', label: '高 (High)', color: '#F59E0B' },
  { value: 'medium', label: '中 (Medium)', color: '#3B82F6' },
  { value: 'low', label: '低 (Low)', color: '#6B7280' },
]

export enum ConditionOperatorEnum {
  EQUALS = 'equals',
  NOT_EQUALS = 'not_equals',
  CONTAINS = 'contains',
  NOT_CONTAINS = 'not_contains',
  IN = 'in',
  NOT_IN = 'not_in',
  GREATER_THAN = 'greater_than',
  LESS_THAN = 'less_than',
  IS_EMPTY = 'is_empty',
  IS_NOT_EMPTY = 'is_not_empty'
}

export const ConditionOperatorOptions = [
  { value: 'equals', label: '等于' },
  { value: 'not_equals', label: '不等于' },
  { value: 'contains', label: '包含' },
  { value: 'not_contains', label: '不包含' },
  { value: 'in', label: '在列表中' },
  { value: 'not_in', label: '不在列表中' },
  { value: 'is_empty', label: '为空' },
  { value: 'is_not_empty', label: '不为空' },
]

export enum ActionTypeEnum {
  CHANGE_STATE = 'change_state',
  SET_PRIORITY = 'set_priority',
  ASSIGN_TO = 'assign_to',
  UNASSIGN = 'unassign',
  ADD_COMMENT = 'add_comment',
  SET_FIELD = 'set_field',
  ARCHIVE = 'archive',
  CLOSE = 'close',
  DISPATCH_AGENT = 'dispatch_agent',
  CALL_WEBHOOK = 'call_webhook',
}

export const ActionTypeOptions = [
  { value: 'change_state', label: '更新状态', icon: '🔄' },
  { value: 'set_priority', label: '设置优先级', icon: '⚡' },
  { value: 'assign_to', label: '分配给', icon: '👤' },
  { value: 'unassign', label: '取消分配', icon: '👥' },
  { value: 'add_comment', label: '添加评论', icon: '💬' },
  { value: 'set_field', label: '设置字段', icon: '📝' },
  { value: 'dispatch_agent', label: '调度 Agent', icon: '🤖' },
  { value: 'call_webhook', label: '调用 Webhook', icon: '🔗' },
]

// ==================== State Transition ====================

export interface StateTransition {
  id: number
  name: string
  description?: string
  source_state_id: number
  target_state_id: number
  issue_type_id?: number
  is_auto: boolean
  project_id: number
  workspace_id: number
  is_deleted: boolean
  created_at: string
  updated_at: string
}

export interface StateTransitionCreate {
  name: string
  description?: string
  source_state_id: number
  target_state_id: number
  issue_type_id?: number
  is_auto?: boolean
}

export interface StateTransitionUpdate {
  name?: string
  description?: string
  is_auto?: boolean
  issue_type_id?: number
}

// ==================== Automation Rule ====================

export interface Trigger {
  type: TriggerTypeEnum
  state_group?: string
  priority?: string
  days_before?: number
  field?: string
}

export interface Condition {
  field: string
  operator: ConditionOperatorEnum
  value?: any
}

export interface Action {
  type: string
  field?: string
  value?: any
}

export interface AutomationRule {
  id: number
  name: string
  description?: string
  is_enabled: boolean
  trigger_type: string
  conditions: string
  actions: string
  execution_count: number
  last_executed_at?: string
  project_id: number
  workspace_id: number
  is_inherited?: boolean
  scope?: string
  schedule_config?: string
  last_triggered_at?: string
  is_deleted: boolean
  created_at: string
  updated_at: string
}

export interface AutomationRuleCreate {
  name: string
  description?: string
  is_enabled?: boolean
  trigger_type: string
  conditions?: string
  actions?: string
  scope?: string
  schedule_config?: string
}

export interface AutomationRuleUpdate {
  name?: string
  description?: string
  is_enabled?: boolean
  trigger_type?: string
  conditions?: string
  actions?: string
  scope?: string
  schedule_config?: string
}

export interface AutomationRuleLite {
  id: number
  name: string
  description?: string
  is_enabled: boolean
  trigger: Trigger
  execution_count: number
}

// ==================== Automation Execution Log ====================

export interface AutomationExecutionLog {
  id: number
  rule_id: number
  status: 'success' | 'failed' | 'skipped'
  trigger_event: string
  triggered_issue_id?: number
  execution_details: Record<string, any>
  error_message?: string
  execution_time_ms?: number
  created_at: string
}

// ==================== Automation Template ====================

export interface AutomationTemplate {
  id: number
  name: string
  description?: string
  category: string
  trigger: Trigger
  conditions: Condition[]
  actions: Action[]
  is_system: boolean
  usage_count: number
}

// ==================== Helper Functions ====================

/**
 * 获取触发器类型的显示名称
 */
export function getTriggerDisplayName(triggerType: TriggerTypeEnum): string {
  const names: Record<TriggerTypeEnum, string> = {
    [TriggerTypeEnum.ISSUE_CREATED]: '工作项创建时',
    [TriggerTypeEnum.ISSUE_UPDATED]: '工作项更新时',
    [TriggerTypeEnum.ISSUE_DELETED]: '工作项删除时',
    [TriggerTypeEnum.ISSUE_ASSIGNED]: '工作项分配时',
    [TriggerTypeEnum.STATE_CHANGED]: '状态变更时',
    [TriggerTypeEnum.DUE_SOON]: '截止日期临近时',
    [TriggerTypeEnum.DUE_DATE_PASSED]: '截止日期过期时',
    [TriggerTypeEnum.CYCLE_STARTED]: '周期开始时',
    [TriggerTypeEnum.CYCLE_ENDED]: '周期结束时',
    [TriggerTypeEnum.COMMENT_ADDED]: '添加评论时',
    [TriggerTypeEnum.SCHEDULED]: '定时执行'
  }
  return names[triggerType] || triggerType
}

/**
 * 获取动作类型的显示名称
 */
export function getActionDisplayName(actionType: ActionTypeEnum): string {
  const names: Record<ActionTypeEnum, string> = {
    [ActionTypeEnum.CHANGE_STATE]: '改变状态',
    [ActionTypeEnum.SET_PRIORITY]: '设置优先级',
    [ActionTypeEnum.ASSIGN_TO]: '分配工作项',
    [ActionTypeEnum.UNASSIGN]: '取消分配',
    [ActionTypeEnum.ADD_COMMENT]: '添加评论',
    [ActionTypeEnum.SET_FIELD]: '设置字段',
    [ActionTypeEnum.ARCHIVE]: '归档',
    [ActionTypeEnum.CLOSE]: '关闭',
    [ActionTypeEnum.DISPATCH_AGENT]: '调度 Agent',
    [ActionTypeEnum.CALL_WEBHOOK]: '调用 Webhook',
  }
  return names[actionType] || actionType
}

/**
 * 获取条件操作符的显示名称
 */
export function getOperatorDisplayName(operator: ConditionOperatorEnum): string {
  const names: Record<ConditionOperatorEnum, string> = {
    [ConditionOperatorEnum.EQUALS]: '等于',
    [ConditionOperatorEnum.NOT_EQUALS]: '不等于',
    [ConditionOperatorEnum.CONTAINS]: '包含',
    [ConditionOperatorEnum.NOT_CONTAINS]: '不包含',
    [ConditionOperatorEnum.IN]: '在列表中',
    [ConditionOperatorEnum.NOT_IN]: '不在列表中',
    [ConditionOperatorEnum.GREATER_THAN]: '大于',
    [ConditionOperatorEnum.LESS_THAN]: '小于',
    [ConditionOperatorEnum.IS_EMPTY]: '为空',
    [ConditionOperatorEnum.IS_NOT_EMPTY]: '不为空'
  }
  return names[operator] || operator
}

/**
 * 获取触发器类型的图标
 */
export function getTriggerIcon(triggerType: TriggerTypeEnum): string {
  const icons: Record<TriggerTypeEnum, string> = {
    [TriggerTypeEnum.ISSUE_CREATED]: 'plus-circle',
    [TriggerTypeEnum.ISSUE_UPDATED]: 'edit',
    [TriggerTypeEnum.ISSUE_DELETED]: 'trash',
    [TriggerTypeEnum.ISSUE_ASSIGNED]: 'user',
    [TriggerTypeEnum.STATE_CHANGED]: 'git-branch',
    
    [TriggerTypeEnum.DUE_SOON]: 'clock',
    [TriggerTypeEnum.DUE_DATE_PASSED]: 'alert-circle',
    [TriggerTypeEnum.CYCLE_STARTED]: 'play',
    [TriggerTypeEnum.CYCLE_ENDED]: 'check-circle',
    [TriggerTypeEnum.COMMENT_ADDED]: 'message-circle',
    [TriggerTypeEnum.SCHEDULED]: 'calendar'
  }
  return icons[triggerType] || 'zap'
}

/**
 * 获取动作类型的图标
 */
export function getActionIcon(actionType: ActionTypeEnum): string {
  const icons: Record<ActionTypeEnum, string> = {
    [ActionTypeEnum.CHANGE_STATE]: 'git-branch',
    [ActionTypeEnum.SET_PRIORITY]: 'flag',
    [ActionTypeEnum.ASSIGN_TO]: 'user-plus',
    [ActionTypeEnum.UNASSIGN]: 'user-minus',
    [ActionTypeEnum.ADD_COMMENT]: 'message-circle',
    [ActionTypeEnum.SET_FIELD]: 'edit-2',
    [ActionTypeEnum.ARCHIVE]: 'archive',
    [ActionTypeEnum.CLOSE]: 'x-circle',
    [ActionTypeEnum.DISPATCH_AGENT]: 'zap',
    [ActionTypeEnum.CALL_WEBHOOK]: 'link',
  }
  return icons[actionType] || 'zap'
}